lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp

```diff
@@ -8588,6 +8588,38 @@ static void tryToElideArgumentCopy(
   }
 }
 
+static bool IsAMDGPUKernel(const Function &F) {
+  return F.getCallingConv() == CallingConv::AMDGPU_KERNEL;
+}
+
+static SmallVector<unsigned, 4> OriginalAlignments(
+  const Function &F,
+  const TargetLowering &TLI,
+  const DataLayout &DL,
+  Type *Ty,
+  SmallVector<unsigned, 4> Align = {}) {
+  if (StructType *STy = dyn_cast<StructType>(Ty)) {
+    unsigned first = Align.size();
+    const StructLayout *SL = DL.getStructLayout(STy);
+
+    for (auto &&E : STy->elements()) {
+      Align = OriginalAlignments(F, TLI, DL, E, std::move(Align));
+    }
+    if (IsAMDGPUKernel(F)) Align[first] = SL->getAlignment();
+  }
+  else if (ArrayType *ATy = dyn_cast<ArrayType>(Ty)) {
+    Align.insert(
+      Align.end(),
+      ATy->getArrayNumElements(),
+      TLI.getABIAlignmentForCallingConv(ATy->getArrayElementType(), DL));
```

Comment:
Do we need to do something with an array of an aggregated type since potentially it may have the same issue?